feat(native): TVF implementation

aditi-pandit opened the pull request:
## Description
Add Table Functions implementation for Prestissimo

## Motivation and Context
RFC-0020-tvf <https://github.com/prestodb/rfcs/blob/main/RFC-0020-tvf.md>


## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

General Changes
* Add Table Functions implementation for Prestissimo

czentgr commented on presto-main-base/src/test/java/com/facebook/presto/metadata/TestTableFunctionRegistry.java:
```diff
@@ -64,19 +65,19 @@ public void testTableFunctionRegistry()
+        assertNotNull(testFunctionRegistry.resolve(SESSION, QualifiedName.of(TEST_FUNCTION)));
+        assertNotNull(testFunctionRegistry.resolve(SESSION, QualifiedName.of(TEST_FUNCTION_2)));
+        assertNotNull(testFunctionRegistry.resolve(SESSION, QualifiedName.of("none")));
+        assertNotNull(testFunctionRegistry.resolve(MISMATCH_SESSION, QualifiedName.of("none")));
```
Previous `assertFalse` should be converted to `assertNull` as suggested by copilot.

czentgr commented on presto-native-execution/presto_cpp/main/connectors/SystemConnector.h:
```diff
@@ -197,4 +197,26 @@ class SystemPrestoToVeloxConnector final : public PrestoToVeloxConnector {
       const final;
 };
 
+class TvfNativePrestoToVeloxConnector final : public PrestoToVeloxConnector {
```
Maybe should be in its own file?

aditi-pandit replied:
same as the .cpp remark..

czentgr commented on presto-native-execution/presto_cpp/main/tvf/exec/TableFunctionPartition.cpp:
```diff
@@ -0,0 +1,415 @@
+  }
+
+  VELOX_CHECK_GT(numRowsLeft, 0);
+  std::vector<int> inputNonNullRows;
```
column_index_t?

aditi-pandit replied:
This should be vector_size_t

czentgr commented on presto-native-execution/presto_cpp/main/PrestoServer.cpp:
```diff
@@ -535,6 +537,53 @@ void PrestoServer::registerHttpEndpoints() {
+              getAnalyzedTableValueFunction(
+                  util::extractMessageBody(body),
+                  server->nativeWorkerPool_.get()));
+        } catch (const velox::VeloxUserError& ex) {
```
Functionally these try-catch do the same thing. 
If we are expecting to catch any exception we can catch std::exception only. They also catch VeloxUserError and VeloxException. Unless catching them would change the response which it does, it just seems added code...

czentgr reviewed (commented):
Thanks! Only a few minor issues. Copilot did also point out a few thing which look legit.

pdabre12 committed:
feat(native) : Add Exclude Columns Tests

pdabre12 committed:
feat(native) : Add TVF C++ Implementation

pdabre12 committed:
feat(native) : Add TVF Native Plugin

pdabre12 committed:
feat(native) : Add dynamic native TVF tests

aditi-pandit commented on presto-native-execution/presto_cpp/main/tvf/exec/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# target_include_directories(presto_tvf_exec PRIVATE
```
Done

aditi-pandit commented on presto-native-execution/presto_cpp/main/tvf/exec/LeafTableFunctionOperator.cpp:
```diff
@@ -0,0 +1,130 @@
+  VELOX_CHECK(
+      result->state() == TableFunctionResult::TableFunctionState::kProcessed);
+  auto resultRows = result->result();
+  VELOX_CHECK(resultRows);
```
This can be null. Have fixed the code.

aditi-pandit commented on presto-native-execution/presto_cpp/main/tvf/core/TableFunctionProcessorNode.cpp:
```diff
@@ -0,0 +1,288 @@
+      sources_.size(), 1, "Number of sources must be equal to 0 or 1");
+}
+
+namespace {
```
This may not be a rule as such, but the anonymous namespaces have the helper functions closest to their use. Say in PlanNode, functions used in the serialization/deserialization of individual PlanNodes were written closest to their implementations.

aditi-pandit commented:
Will add TVF documentation subsequently.

aditi-pandit committed:
refactor: Fix checkstyle warnings

aditi-pandit committed:
fix(docs): Add C++ Table functions documentation

aditi-pandit commented:
@czentgr : Please can you do a round of review. I fixed the failing tests, added design/SPI doc and addressed previous review comments

@mohsaka @pdabre12 @xin-zhang2 : PTAL also

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+======================================================
```
```suggestion
=============================================
```

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+======================================================
+Table-Valued Functions (TVF) — C++ API Design
+======================================================
```
```suggestion
=============================================
```

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+======================================================
+Table-Valued Functions (TVF) — C++ API Design
```
Could this be changed to be consistent with the other sub-page titles in https://prestodb.io/docs/current/presto-cpp.html? 

Maybe, looking at your description of this page in line 10
`This document covers the **C++ native worker Table Function SPI** and provides worked examples for it.`

something like
`Presto C++ Table Function SPI`

Note: if you change this line, please edit lines 1 and 3 to be the same length.

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+Argument Types
+--------------
+
++------------+--------------------------------------------------+------------------------------------------------------+
```
The Argument Types and Table Argument Modifiers tables are missing in a local doc build. See screenshot. 
The spacing in lines 35 and 37 in `Argument Types` look inconsistent. 
The spacing in lines 52, 56, and 58 in `Table Argument Modifiers` look inconsistent. 

<img width="704" height="464" alt="Image" src="https://github.com/user-attachments/assets/5b453202-64c9-426f-9698-6ec9bf4a340d" />

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+Parsing
+-------
+
+The SQL parser recognises the TABLE(…) wrapper in the FROM clause and builds an
```
```suggestion
The SQL parser recognises the ``TABLE(…)`` wrapper in the ``FROM`` clause and builds an
```

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+
+The SQL parser recognises the TABLE(…) wrapper in the FROM clause and builds an
+AstNode for the table function invocation. Each argument is categorised by its
+syntactic form — TABLE(subquery) becomes a table argument, DESCRIPTOR(…) becomes a
```
```suggestion
syntactic form — ``TABLE(subquery)`` becomes a table argument, ``DESCRIPTOR(…)`` becomes a
```

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+The SQL parser recognises the TABLE(…) wrapper in the FROM clause and builds an
+AstNode for the table function invocation. Each argument is categorised by its
+syntactic form — TABLE(subquery) becomes a table argument, DESCRIPTOR(…) becomes a
+descriptor argument, and typed constants become scalar arguments. PARTITION BY,
```
```suggestion
descriptor argument, and typed constants become scalar arguments. ``PARTITION BY``,
```

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+AstNode for the table function invocation. Each argument is categorised by its
+syntactic form — TABLE(subquery) becomes a table argument, DESCRIPTOR(…) becomes a
+descriptor argument, and typed constants become scalar arguments. PARTITION BY,
+ORDER BY, and COPARTITION clauses attached to table arguments are also captured at
```
```suggestion
``ORDER BY``, and ``COPARTITION`` clauses attached to table arguments are also captured at
```

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+
+* **ScalarArgument** — Comprises the constant value and its Velox type.
+* **TableArgument** — Consists of the row type of the input relation, plus any partition/sort keys the caller specified.
+* **DescriptorArgument** — has the field names (and optional types) from the DESCRIPTOR(…) in the table function invocation.
```
```suggestion
* **DescriptorArgument** — has the field names (and optional types) from the ``DESCRIPTOR(…)`` in the table function invocation.
```

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+
+The planner then rewrites TableFunctionNode into a PlanNode subtree based on the input shape:
+
++------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------+
```
This table is also missing from the local doc build.

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+--------
+
+Polymorphism for Table Functions
+=================================
```
```suggestion
================================
```

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+This design enables maximum flexibility while keeping the C++ SPI clean and intuitive.
+
+Query Processing Flow for a SQL with a Table Function
+======================================================
```
```suggestion
=====================================================
```

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+responsible for Table function execution. More in the design section.
+
+Split Generation (leaf functions only)
+---------------------------------------
```
```suggestion
--------------------------------------
```

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+exactly like ordinary connector splits.
+
+Serialization and dispatch
+---------------------------
```
```suggestion
--------------------------
```

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+* **TableFunctionOperator**: if the function has input tables
+
+LeafTableFunctionOperator
+^^^^^^^^^^^^^^^^^^^^^^^^^^
```
```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^
```

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+=======
+
+Defining and registering a Table Function
+------------------------------------------
```
```suggestion
-----------------------------------------
```

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+To define and register a Table Function, the user can perform the following steps:
+
+* Identify the Table function arguments. These are represented with ArgumentSpecification classes when registering the function, but are resolved to Argument classes when instantiated in a SQL query.
+* Identify the Table function return type. The Table function returns a table though its columns could be generic (decided at analysis time), described (known at specification time) or PassThrough (no new columns, the input table is passthrough)
```
```suggestion
* Identify the Table function return type. The Table function returns a table though its columns could be
  * generic (decided at analysis time)
  * described (known at specification time)
  * PassThrough (no new columns, the input table is passthrough)
```

Maybe separate the three options here into an unordered list, so the reader can find them easily?

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+Each section below goes over the SPI for each of these parts.
+
+Argument and ArgumentSpecification Classes
+-------------------------------------------
```
```suggestion
------------------------------------------
```

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+   };
+
+1b. ScalarArgumentSpecification / ScalarArgument
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+   };
+
+1c. TableArgumentSpecification / TableArgument
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+   };
+
+1d. DescriptorArgumentSpecification / DescriptorArgument
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+--------
+
+Return Type Specification
+--------------------------
```
```suggestion
-------------------------
```

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+Return Type Specification
+--------------------------
+
+Specifies the Table Function Return type. This could be GENERIC (so determined at Analysis time), DESCRIBED (so fixed
```
Same suggestion as earlier about an unordered list. 

Also, the earlier text had the options as 
```
generic (decided at analysis time), described (known at specification time) or PassThrough (no new columns, the input table is passthrough)
```
Suggest choosing one format (example: generic, GENERIC) and description for each, and using the same words in both places.

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+Table Functions are processed as TableFunctionDataProcessor(s) or TableFunctionSplitProcessor(s)
+
+TableFunctionDataProcessor
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^^
```

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+Execution Processors
+--------------------
+
+Table Functions are processed as TableFunctionDataProcessor(s) or TableFunctionSplitProcessor(s)
```
```suggestion
Table Functions are processed as TableFunctionDataProcessor(s) or TableFunctionSplitProcessor(s).
```

Do these two items need to be formatted as in https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines#formatting? 
"Use verbatim monospace blocks for types, SQL keywords and configuration properties:"

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,532 @@
+Table-Valued Functions (TVF) — C++ API Design
+======================================================
+
+Overview
```
```suggestion
.. contents::
    :local:
    :backlinks: none
    :depth: 1

Overview
```
This page has many headings. Consider including a table of contents within the page.

steveburnett reviewed (changes requested):
Thank you for all of this valuable documentation! Most of my comments are formatting issues. 

Please add this page to 
https://github.com/prestodb/presto/blob/master/presto-docs/src/main/sphinx/presto-cpp.rst

so that it displays in the table of contents. 

Add 
`    presto_cpp/table-functions`

jja725 commented on presto-native-execution/presto_cpp/main/tvf/exec/LeafTableFunctionOperator.cpp:
```diff
@@ -0,0 +1,127 @@
+        blockingFuture_);
+
+    if (blockingReason_ != BlockingReason::kNotBlocked) {
+      return nullptr;
```
not too sure about the semantic but is it expected to have nullptr here instead of a blocking future?

aditi-pandit replied:
Thanks for catching this. Have fixed the blocking logic to be split across the split blocking and the function blocking

jja725 commented on presto-native-execution/presto_cpp/main/tvf/spi/TableFunction.cpp:
```diff
@@ -0,0 +1,130 @@
+    TableFunctionSplitProcessorFactory splitProcessorfactory,
+    TableFunctionSplitGenerator splitGenerator) {
+  auto sanitizedName = exec::sanitizeName(name);
+  tableFunctions().insert(
```
insert is a no-op on duplicate keys, which is different from the function doc comment, should we use insert_or_assign if it's expected to override

jja725 commented on presto-native-execution/presto_cpp/main/tvf/spi/TableFunction.h:
```diff
@@ -0,0 +1,248 @@
+      const std::string& name,
+      velox::memory::MemoryPool* pool,
+      velox::HashStringAllocator* stringAllocator)
+      : pool_(pool), stringAllocator_(stringAllocator) {}
```
name is dropped

aditi-pandit replied:
Fixed

jja725 commented on presto-native-execution/presto_cpp/main/functions/FunctionMetadata.cpp:
```diff
@@ -267,6 +271,115 @@ json buildWindowMetadata(
+      descriptorArgumentSpecification->name = descriptorArgumentSpec->name();
+      descriptorArgumentSpecification->defaultValue =
+          buildDescriptor(descriptorArgumentSpec->descriptor());
+      descriptorArgumentSpecification->required = false;
```
is this hardcode false expected here?  do we want to propagate descriptorArgumentSpec->required()

aditi-pandit replied:
Thanks for catching this.

jja725 commented on presto-main-base/src/main/java/com/facebook/presto/metadata/TableFunctionRegistry.java:
```diff
@@ -112,7 +113,7 @@ public static List<CatalogSchemaFunctionName> toPath(Session session, QualifiedN
      * Table functions are resolved case-insensitive for consistency with existing scalar function resolution.
      */
-    public Optional<TableFunctionMetadata> resolve(Session session, QualifiedName qualifiedName)
+    public TableFunctionMetadata resolve(Session session, QualifiedName qualifiedName)
```
could you help me understand more why we need this return type change in java?

aditi-pandit replied:
This really doesn't have much consequence, just made some code easier to write. This isn't an external API so its okay I think.

jja725 commented on presto-native-execution/presto_cpp/main/tvf/exec/LeafTableFunctionOperator.h:
```diff
@@ -0,0 +1,89 @@
+
+  void reclaim(
+      uint64_t targetBytes,
+      velox::memory::MemoryReclaimer::Stats& stats) override;
```
these operator potentially can implement reclaim, maybe add to TODO comment

aditi-pandit replied:
Done

jja725 commented on presto-native-tvf/src/checkstyle/presto-checks.xml:
```diff
@@ -0,0 +1,219 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
is there any blocker to remove this style exception case and fix the style in these java code

aditi-pandit replied:
@pdabre12 : PTAL.

pdabre12 replied:
Don't know if we need this, can you try removing it? There's a check style file already in `presto-root/`

aditi-pandit commented:
> Thanks for this great contribution, looking forward to use it for vector search. BTW, have we discussed with velox community about this? I feel the table function operator part can just be in velox. Trying to review this huge pr more but leave a round of comments first.

@jja725 : Thanks for your review. Like we discussed we will put this proposal to the Velox team after evaluating if it can be used in Spark as well.

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/table-functions.rst:
```diff
@@ -0,0 +1,540 @@
+
+The planner then rewrites TableFunctionNode into a PlanNode subtree based on the input shape:
+
++------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------+
```
Please fix the formatting of this table. It is not being compiled. See screenshot of a local doc build. 

<img width="691" height="345" alt="Screenshot 2026-08-13 at 10 34 16 AM" src="https://github.com/user-attachments/assets/32184e5f-737b-4e7a-b03e-5e612424ea54" />

aditi-pandit replied:
Thanks @steveburnett. Have been able to verify with the local build now. PTAL.

steveburnett replied:
> Thanks @steveburnett. Have been able to verify with the local build now. PTAL.

Yes, I see it now in a new build. Thank you!

aditi-pandit commented on presto-native-execution/presto_cpp/main/connectors/SystemConnector.cpp:
```diff
@@ -394,4 +395,49 @@ std::unique_ptr<protocol::ConnectorProtocol>
 }
+
+std::unique_ptr<velox::connector::ConnectorSplit>
+TvfNativePrestoToVeloxConnector::toVeloxSplit(
```
The TVF translation is only needed for splits. We attached all tvfs to system connector so the columns handles and table handles are transformed to SystemTableHandle/SystemColumnHandle. Hence, this code is in this class.

aditi-pandit commented on presto-main-base/src/main/java/com/facebook/presto/connector/system/SystemConnectorModule.java:
```diff
@@ -80,6 +83,9 @@ public void configure(Binder binder)
 
         binder.bind(GlobalSystemConnectorFactory.class).in(Scopes.SINGLETON);
         binder.bind(SystemConnectorRegistrar.class).asEagerSingleton();
+        binder.bind(PluginNodeManager.class).in(Scopes.SINGLETON);
+        binder.bind(NodeManager.class).to(PluginNodeManager.class).in(Scopes.SINGLETON);
+        binder.bind(FunctionAndTypeManager.class).in(Scopes.SINGLETON);
```
@pdabre12 : Any comments ?

pdabre12 replied:
Yes, it seems valid, pls remove it.
